lib/vm/src/memory.rs

```diff
@@ -12,30 +12,38 @@ use wasmer_runtime_core::{
 /// This is the same as cosmwasm::memory::Slice
 /// but defined here to allow wasm impl
 #[repr(C)]
-#[derive(Default, Clone, Copy)]
+#[derive(Default, Clone, Copy, Debug)]
 pub struct Slice {
     pub offset: u32,
     pub len: u32,
 }
 
 unsafe impl ValueType for Slice {}
 
+// Expects a (fixed size) Slice struct at ptr, which is read. This links to the
+// memory region, which is read in the second step.
 pub fn read_memory(ctx: &Ctx, ptr: u32) -> Vec<u8> {
     let slice = to_slice(ctx, ptr);
     let memory = ctx.memory(0);
-    let len = slice.len as usize;
-    let mut result = vec![0u8; len];
 
     // TODO: there must be a faster way to copy memory
-    let buffer = WasmPtr::<u8, Array>::new(slice.offset)
-        .deref(memory, 0, slice.len)
-        .unwrap();
-    for i in 0..len {
-        // result[i] = unsafe { buffer.get_unchecked(i).get() }
-        // resolved to memcpy, but only if we really start copying huge arrays
-        result[i] = buffer[i].get();
+    match WasmPtr::<u8, Array>::new(slice.offset).deref(memory, 0, slice.len) {
```

Comment:
Interesting cleanup to handle invalid pointers. Want to make a PR on master with just this change (and maybe the vm/testing one as well?)

Reply:
i.e. you don't want to merge this commit https://github.com/confio/cosmwasm/pull/118/commits/7f2d31047f446bcc8acf82a5e70539694185c92b without the rest of the integration finished? In this case I'd just remove this commit from here.

Reply:
Nevermind, I create a new PR avoiving the term "AssemblyScript" for clarity

Reply:
#119 119